apps/generator/lib/generator.js

```diff
@@ -867,34 +866,58 @@ class Generator {
     const relativeTargetFile = path.relative(this.targetDir, targetFile);
-
+    let shouldGenerate = true;
+  
     if (shouldIgnoreFile(relativeSourceFile)) return;
-
-    const shouldOverwriteFile = await this.shouldOverwriteFile(relativeTargetFile);
-    if (!shouldOverwriteFile) return;
-
-    if (this.templateConfig.conditionalFiles?.[relativeSourceFile]) {
-      const server = this.templateParams.server && asyncapiDocument.servers().get(this.templateParams.server);
-      const source = jmespath.search({
-        ...asyncapiDocument.json(),
-        ...{
-          server: server ? server.json() : undefined,
-        },
-      }, this.templateConfig.conditionalFiles[relativeSourceFile].subject);
-
-      if (!source) return log.debug(logMessage.relativeSourceFileNotGenerated(relativeSourceFile, this.templateConfig.conditionalFiles[relativeSourceFile].subject));
-
-      if (source) {
-        const validate = this.templateConfig.conditionalFiles[relativeSourceFile].validate;
-        const valid = validate(source);
-        if (!valid) return log.debug(logMessage.conditionalFilesMatched(relativeSourceFile));
-      }
+    
+    if (!(await this.shouldOverwriteFile(relativeTargetFile))) return;
+
+    let conditionalPath = '';
+    if (
+      this.templateConfig.conditionalFiles &&
+      this.templateConfig.conditionalGeneration
+    ) {
+      log.debug(
+        'Both \'conditionalFiles\' and \'conditionalGeneration\' are defined. Ignoring \'conditionalFiles\' and using \'conditionalGeneration\' only.'
+      );
+    }
+    
+    if (this.templateConfig.conditionalGeneration?.[relativeSourceDirectory]) {
+      conditionalPath = relativeSourceDirectory;
+    } else if (this.templateConfig.conditionalGeneration?.[relativeSourceFile]) {
+      conditionalPath = relativeSourceFile;
+    } else
+    if (this.templateConfig.conditionalFiles?.[relativeSourceFile]) {  
+      // conditionalFiles becomes deprecated with this PR, and soon will be removed.
+      // TODO: https://github.com/asyncapi/generator/issues/1553
+      conditionalPath = relativeSourceDirectory;
+    }
+    
+    if (conditionalPath) {
+      shouldGenerate = await conditionalGeneration(
+        relativeSourceFile,
+        relativeSourceDirectory,
+        relativeTargetFile,
+        this.templateConfig,
+        this.targetDir,
+        conditionalPath,
+        this.templateParams,
+        asyncapiDocument
+      );
+    }
+  
+    if (shouldGenerate) {
+      if (this.isNonRenderableFile(relativeSourceFile)) {
+        log.debug(`${relativeSourceFile} is a non-renderable file and was copied without changes.`);
+        return await copyFile(sourceFile, targetFile); 
+      } 
+      log.debug(`Successfully rendered template and wrote file ${relativeSourceFile} to location: ${targetFile}`);
+      await this.renderAndWriteToFile(asyncapiDocument, sourceFile, targetFile);
```

Comment:
I don't fully understand this, we log success before rendering? and if `shouldGenerate` is true, we generate

it's a bit confusing


maybe the most confusing is function name `conditionalGeneration` and current definition that says that if `shouldGenerate` is true then generation should run, not the other way around

Reply:
I changed the jsdoc for proper understanding thanks @derberg